unsaved_fast_iterator.go

```diff
@@ -0,0 +1,231 @@
+		start:        start,
+		end:          end,
+		err:          nil,
+		ascending:    ascending,
+		ndb:          ndb,
+		unsavedFastNodeAdditions: unsavedFastNodeAdditions,
+		unsavedFastNodeRemovals: unsavedFastNodeRemovals,
+		unsavedFastNodesToSort: make([]string, 0),
+		nextKey: nil,
+		nextVal: nil,
+		nextUnsavedNodeIdx: 0,
+		fastIterator: NewFastIterator(start, end, ascending, ndb),
+	}
+
+	// We need to ensure that we iterate over saved and unsaved state in order.
+	// The strategy is to sort unsaved nodes, the fast node on disk are already sorted.
+	// Then, we keep a pointer to both the unsaved and saved nodes, and iterate over them in order efficiently.
+	for _, fastNode := range unsavedFastNodeAdditions {
+		if start != nil && bytes.Compare(fastNode.key, start) < 0 {
+			continue
+		}
+
+		if end != nil && bytes.Compare(fastNode.key, end) >= 0 {
+			continue
+		}
+
+		iter.unsavedFastNodesToSort = append(iter.unsavedFastNodesToSort, string(fastNode.key))
+	}
+
+	sort.Slice(iter.unsavedFastNodesToSort, func(i, j int) bool {
+		if ascending{
+			return iter.unsavedFastNodesToSort[i] < iter.unsavedFastNodesToSort[j]
+		} else {
+			return iter.unsavedFastNodesToSort[i] > iter.unsavedFastNodesToSort[j]
+		}
+	})
+
+	if iter.ndb == nil {
+		iter.err = errFastIteratorNilNdbGiven
+		iter.valid = false
+		return iter
+	}
+
+	if iter.unsavedFastNodeAdditions == nil {
+		iter.err = errUnsavedFastIteratorNilAdditionsGiven
+		iter.valid = false
+		return iter
+	}
+
+	if iter.unsavedFastNodeRemovals == nil {
+		iter.err = errUnsavedFastIteratorNilRemovalsGiven
+		iter.valid = false
+		return iter
+	}
+
+	// Move to the first elemenet
+	iter.Next()
+
+	return iter
+}
+
+// Domain implements dbm.Iterator.
+// Maps the underlying nodedb iterator domain, to the 'logical' keys involved.
+func (iter *UnsavedFastIterator) Domain() ([]byte, []byte) {
+	return iter.start, iter.end
+}
+
+// Valid implements dbm.Iterator.
+func (iter *UnsavedFastIterator) Valid() bool {
+	if iter.start != nil && iter.end != nil {
+		if bytes.Compare(iter.end, iter.start) != 1 {
+			return false
+		}
+	}
+
+	return iter.fastIterator.Valid() || iter.nextUnsavedNodeIdx < len(iter.unsavedFastNodesToSort) || iter.nextKey != nil || iter.nextVal != nil
+}
+
+// Key implements dbm.Iterator
+func (iter *UnsavedFastIterator) Key() []byte {
+	return iter.nextKey
+}
+
+// Value implements dbm.Iterator
+func (iter *UnsavedFastIterator) Value() []byte {
+	return iter.nextVal
+}
+
+// Next implements dbm.Iterator
+func (iter *UnsavedFastIterator) Next() {
+	if iter.ndb == nil {
+		iter.err = errFastIteratorNilNdbGiven
+		iter.valid = false
+		return
+	}
+
+	if iter.fastIterator.Valid() && iter.nextUnsavedNodeIdx < len(iter.unsavedFastNodesToSort) {
+		diskKeyStr := string(iter.fastIterator.Key())
+
+		if iter.unsavedFastNodeRemovals[diskKeyStr] != nil {
+			// If next fast node from disk is to be removed, skip it.
+			iter.fastIterator.Next()
+			iter.Next()
+			return
```

Comment:
Spent a silly amount of time looking into the golang compiler looking into how long an empty hashmap lookup takes. End result: I can't figure out how the map source code works, but I'm sure its blazingly fast and its not a concern lol.